src/reprepro_updater/helpers.py

```diff
@@ -142,11 +159,38 @@ def invalidate_dependent(repo_dir, distro, arch, package):
     """
     Remove all dependents of the package with the same arch.
 
-    This is only valid for binary packages.
+    This is only valid for binary packages and assumes all
+    packages are in the `main` component.
     """
-    # storage for recursion
-    processed_packages = []
-    return _invalidate_dependent(repo_dir, distro, arch, package, processed_packages)
+    # Use internal repository information parser to get reverse dependents.
+    # Using reprepro for each package was resulting in huge runtimes causing timeouts
+    # on the buildfarm when invalidating low-level packages.
+    repo_info = RepositoryInfo(repo_dir, distro, arch)
+
+    # queue of dependents to iterate
+    dependents_to_process = repo_info.get_rdepends(package)
```

Comment:
Since this line doesn't copy the returned set it will mutate the data later on (calling `pop()`) which is likely not desired.

Reply:
This is a good point and is true of anything that gets a set from get_rdepends. I think It's best resolved by copying the memoized set when returning it rather than making each caller remember to do so.

Reply:
The returned set is now a copy of the internal one as of 9839c51